### ISSUE TITLE
Deprecate connectWith and listenWith

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6958]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)| @moshez reported|
|-|-|
|Trac ID|trac#367|
|Type|enhancement|
|Created|2003-10-24 13:18:18Z|
|Branch|https://github.com/twisted/twisted/tree/ireactorarbitrary-deprecation-367-2|

<details><summary>Searchable metadata</summary>

```
trac-id__367 367
type__enhancement enhancement
reporter__moshez moshez
priority__high high
milestone__ 
branch__branches_ireactorarbitrary_deprecation_367_2 branches/ireactorarbitrary-deprecation-367-2
branch_author__exarkun__jesstess exarkun, jesstess
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1067001498000000 1067001498000000
changetime__1267633840000000 1267633840000000
version__ 
owner__ 
cc__moshez cc__thijs cc__exarkun cc__itamarst cc__radix
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
connectWith and listenWith are powerful escape mechanisms
but are poorly documented.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Actually, this interface should probably be deprecated.  It's redundant with `IReactorFDSet`.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra commented|
|-|-|

Replying to [exarkun](#note_4):
> Actually, this interface should probably be deprecated.  It's redundant with `IReactorFDSet`.

Perhaps we should rename this ticket in that case, or close it and open a new one for the deprecated interface?

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra commented|
|-|-|

Renaming ticket.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @jesstess|
|-|-|

Hey, let's do this soon so we can get rid of them.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_8"></a>|@jesstess commented|
|-|-|

(In [28134]) Branching to 'ireactorarbitrary-deprecation-367'

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_9"></a>|@jesstess commented|
|-|-|

(In [28135]) Deprecate IReactorArbitrary and add a unit test for the deprecation in
test_posixbase.

refs [#6958](https://github.com/twisted/twisted/issues/6958)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra commented|
|-|-|

Can you open a ticket for removing these interfaces?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

Unfortunately, due to [#4166](https://github.com/twisted/twisted/issues/4166):

```
>>> from twisted.internet import interfaces
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "twisted/internet/interfaces.py", line 11, in <module>
    from twisted.python.deprecate import deprecatedModuleAttribute
  File "twisted/python/deprecate.py", line 58, in <module>
    from twisted.python.reflect import fullyQualifiedName
  File "twisted/python/reflect.py", line 817, in <module>
    from twisted.python.deprecate import deprecated
ImportError: cannot import name deprecated
>>>
```

I think we need to fix [#4166](https://github.com/twisted/twisted/issues/4166) before we can use `twisted.python.deprecate` in `twisted.internet.interfaces`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set owner to @jesstess|
|-|-|

  1. I think we should avoid the deprecation warning which is currently emitted when `twisted.internet.posixbase` is imported.  How about making a private alias for `IReactorArbitrary` in `interfaces.py` and using that elsewhere internally where we need it?  That should let us continue to implement the interface (necessary for our backwards compatibility policy) but not generate a spurious warning for applications that aren't _using_ our implementation of the interface.
  1. We should, however, emit a deprecation warning if the reactor's `connectWith` or `listenWith` method is actually invoked.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

Oh.  It'd be nice to merge forward too, so that we can get a clean test build without the [#4166](https://github.com/twisted/twisted/issues/4166) problems that are now fixed in trunk.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_15"></a>|@jesstess commented|
|-|-|

(In [28514]) Deprecate connectWith and listenWith in PosixReactorBase. Import an
alias for IReactorArbitrary in posixbase to avoid the deprecation
warning.

refs [#6958](https://github.com/twisted/twisted/issues/6958)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_16"></a>|@jesstess removed owner|
|-|-|

The branch now has what is hopefully a correct interpretation of exarkun's review comments above, and has been merged forward.

Build results [here](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/ireactorarbitrary-deprecation-367)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [28553]) Branching to 'ireactorarbitrary-deprecation-367-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

(In [28554]) Merge forward, resolve posixbase.py conflict, drop the unfortunate r28516

refs [#6958](https://github.com/twisted/twisted/issues/6958)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

I think my review was unclear in a couple places.  Sorry about that.  I'm going to make the changes I had in mind, as well as a couple others, and re-submit for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun commented|
|-|-|

(In [28555]) Just alias the IReactorArbitrary name, don't completely redefine it

refs [#6958](https://github.com/twisted/twisted/issues/6958)

The alias lets posixbase.py continue to import the interface so all the posixbase-derived
reactors can claim to implement it while avoiding getting a deprecation warning spat out
as a result.  It needs to be an alias rather than a duplicate definition so that code that
is inspecting the reactor to see what interfaces it provides can correctly determine whether
`IReactorArbitrary` is implemented, since these checks are done based on the identity of the
interface object in question.

Also change the deprecation warning strings so that they include the version of Twisted in
which this deprecation happened (although the rest of the string still does not quite seem
right).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun commented|
|-|-|

(In [28556]) Update the IReactorArbitrary deprecation message and make the tests match

refs [#6958](https://github.com/twisted/twisted/issues/6958)

`deprecatedModuleAttribute` creates a message which includes a bunch of information already,
so omit the duplicate stuff from the additional message text supplied.  Also bump the version
as we have now missed 10.0 for this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_23"></a>|@therve set owner to @exarkun|
|-|-|

Looks good! One test uses `IReactorArbitrary` in test_application, for no obvious reason. Maybe it can be fixed.

Please merge once fixed, or open a ticket to remove the warning. Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun commented|
|-|-|

(In [28637]) Deprecate GenericServer and GenericClient; suppress the appropriate deprecation warnings in the test suite

refs [#6958](https://github.com/twisted/twisted/issues/6958)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun removed owner|
|-|-|

I added deprecations for `GenericServer` and `GenericClient`, tests for same, and suppressed all these new warnings in `test_everythingThere`.  Also bumped `internet.py`'s copyright date in r28638.  [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/ireactorarbitrary-deprecation-367-2).

Thanks

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_26"></a>|@therve set owner to @exarkun|
|-|-|

Looks great, +1!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun commented|
|-|-|

(In [28640]) Mention GenericServer and GenericClient in the news fragment

refs [#6958](https://github.com/twisted/twisted/issues/6958)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_28"></a>|@exarkun set status to `closed`|
|-|-|

(In [28641]) Merge ireactorarbitrary-deprecation-367-2 

Author: jesstess, exarkun
Reviewer: exarkun, therve
Fixes: [#6958](https://github.com/twisted/twisted/issues/6958)

Deprecate twisted.internet.interfaces.IReactorArbitrary, as well as the
implementation of this interface in `twisted.internet.posixbase`, and
the `IService` implementations which wrapped its methods, `GenericServer`
and `GenericClient` from `twisted.application.internet`.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
